refactor: Improve internal handling of narrow links

Ezio-Sarthak opened the pull request:
<!-- Please see https://github.com/zulip/zulip-terminal#contributor-guidelines ! -->

**What does this PR do?**  <!-- Overall description goes here -->
This PR refactors `MessageLinkButton` in an attempt to improve narrow links' validation, patching, and relavant tests.
<!-- If fixing a filed bug or new feature, add 'Fixes #<issue>' or 'Partial fix for #<issue>' -->

<!-- Add a link to a discussion on chat.zulip.org, if relevant -->

**Tested?** <!-- Fine to leave some of these unchecked if this is a draft/work-in-progress -->
- [x] Manually
- [x] Existing tests (adapted, if necessary)
- [x] New tests added (for any new behavior)
- [x] Passed linting & tests (each commit)
<!-- Code must pass CI (GitHub Actions) before merging - look for the green tick! -->

<!-- See https://github.com/zulip/zulip-terminal#commit-style -->
**Commit flow** <!-- if more than one commit; add/delete/fill-in as appropriate -->
- refactor: tests: Use `case` for patching test ids of large frequency.
- refactor: buttons: Use new helpers to patch and validate data.
- refactor: tests: Improve and add tests for validating narrow link.
- refactor: buttons: Extract topic name validation to separate method.
- refactor: buttons: Improve stream data handling via typed classes.

**Notes & Questions** <!-- if any; add/delete/fill-in as appropriate -->
<!-- For example:
- this doesn't include feature X (yet?)
- unsure about Y
- should this do Z?
-->

**Interactions** <!-- if any; add/delete/fill-in as appropriate -->
<!-- eg.
- Waiting on #<PR>
- Blocks #<PR>
-->

**Visual changes** <!-- if any; add/delete/fill-in with screenshot/diagram as appropriate -->

Ezio-Sarthak committed:
refactor: buttons: Extract patch method from stream validation helper.

This commit:
 - extracts patching of stream data to `_patch_narrow_link`, which would
   potentally be a centralized helper to patch validated data. This way,
   explicit patching of stream/pm data is ensured.
 - extracts validation logic of stream data to `_validate_stream_data`
   and thus slightly tidies the general `_validate_narrow_link`. Since
   this stream helper is being called from `_validate_narrow_link`, the
   relavant tests for it shall stay in the centralized helper itself.

Tests amended.

Ezio-Sarthak committed:
tests: test_buttons: Use `case` for test ids of large frequency.

Ezio-Sarthak committed:
tests: test_buttons: Improve tests for narrow link validation.

This commit improves and makes the tests for `_validate_narrow_link`
more justified. Following changes are done:
 - Tests added for deprecated topic narrow links (both valid and in-
   valid)
 - Parameter `stream_id_from_name` added to parametrize, so as to
   patch the return value in case stream_id isn't provided and the
   stream is a valid stream. (e.g., for valid deprecated links).

A FIXME is added for potential improvement in the test parameters.

Ezio-Sarthak committed:
refactor: buttons: Improve stream data handling via typed classes.

Currently stream narrow data is being primarily used for two purposes:
 - Validating the narrow link,
 - Patching the relevant information for successful narrowing.

The stream data under validation has either only stream id or only
stream name, which makes it a better candidate for `total=False`.
Thus, this type of data it is extracted to a separate TypedDict class
`DecodedStreamPartial`.

Once the stream data is successfully validated, we're ready to patch
the relevant stream data. This should be strictly typed since now we
know that the stream data is valid. Thus, this type of data is
extracted to a separate TypedDict class `DecodedStreamAbsolute`.

Tests amended.

preetmishra commented on zulipterminal/ui_tools/buttons.py:
```diff
@@ -322,14 +322,20 @@ def keypress(self, size: urwid_Size, key: str) -> Optional[str]:
+    stream_name: str
+
+
+class DecodedStreamPatched(TypedDict):
```
How about `DecodedStreamPartial` and `DecodedStreamAbsolute`?

Ezio-Sarthak replied:
Sounds good :+1

preetmishra commented on zulipterminal/ui_tools/buttons.py:
```diff
@@ -507,17 +513,19 @@ def _patch_narrow_link(self, parsed_link: ParsedNarrowLink) -> None:
                 stream_name = cast(str, model.stream_dict[stream_id]["name"])
-                parsed_link["stream"]["stream_name"] = stream_name
+
+            parsed_link["stream_patched"] = DecodedStreamPatched(
```
We could also delete the partial stream data from the `dict` at this point.

preetmishra reviewed (commented):
@Ezio-Sarthak Thanks for taking this up! This would greatly improve the internal-link-handling code. :+1: 

This looks good mostly. I have left in-line comments about how we could the commits coherent and improve the commit flow.

Ezio-Sarthak committed:
refactor: buttons: Extract topic name validation to separate method.

This commit isolates the logic of topic name validation of narrow
links to a new helper `_validate_topic_data`. This makes the overall
flow of validation more clean.

preetmishra commented on zulipterminal/ui_tools/buttons.py:
```diff
@@ -469,6 +469,23 @@ def _validate_stream_data(self, parsed_link: ParsedNarrowLink) -> str:
+        stream_name = parsed_link["stream"]["stream_name"]
+        topic_name = parsed_link["topic_name"]
+
+        if not stream_id:  # Primarily for valid deprecated links
+            stream_id = cast(int, self.model.stream_id_from_name(stream_name))
```
The tests pass even if this block is deleted, which suggests that we should reconsider the approach we discussed.

preetmishra reviewed (commented):
@Ezio-Sarthak Thanks for the updates! :+1: This looks good except for the third commit where we improve the tests. I have left some comments in-line.

Ezio-Sarthak commented on zulipterminal/ui_tools/buttons.py:
```diff
@@ -447,10 +447,10 @@ def _parse_narrow_link(cls, link: str) -> ParsedNarrowLink:
         return parsed_link
 
-    def _validate_and_patch_stream_data(self, parsed_link: ParsedNarrowLink) -> str:
+    def _validate_stream_data(self, parsed_link: ParsedNarrowLink) -> str:
```
Yes, this is exactly the case. I'll mention this in the commit message. Thanks for pointing out :+1:
